### DISCUSSION TITLE
How do I control the setting of the input field state?

### DISCUSSION THREAD

**Discussion by mule-stand** [ariakit/ariakit#3266] [Questions]
I am using useValidate to validate the form.
In my case I want to not only validate the user input, but have access to what the user enters before the actual state change (by suoti mutate it).

The example is this, I have a form with a name value and there is validation in `useValidate` for example:
```ts
formStore.useValidate((state) => {
    const name = state.values.name
    if (name.length < 5) {
      formStore.setError(state.values.name, 'Error message')
    }
  })
  ```

But I, for example, want to disallow the use of the "z" character on user input. Using a normal input and `onChange` in it, I could just check it and if this symbol is present, then do not update the state, so the input would not change.
And after checking if it is less than 5 characters then show an error.

This example is farfetched, it may be a logical question why I don't just do it in `useValidate` and show an error? But this is a simplified example, in my case I have an input field with a transfer amount, which is validated using regular expressions and simply does not allow to enter a non-valid amount. And in `useValidate` I already check, for example, whether this amount is less than 10000 and then I already show an error, say the maximum amount is 10000.
I hope I explained it clearly (don't hesitate to clarify, besides my English is not at a high level).

In the end, three questions:
1. Can I embed myself in the state setting process and somehow mutate it 
or prevent it from changing?
2. Am I thinking right in general, or maybe I'm going down the wrong path?
3. What is the difference between `useValidate` and `onValidate`?

**Comment by diegohaz** [comment]
> 1. Can I embed myself in the state setting process and somehow mutate it
>   or prevent it from changing?
> 2. Am I thinking right in general, or maybe I'm going down the wrong path?

It's hard to understand your use case. I believe it would be beneficial if you could provide a concrete example instead of an abstract one. Screenshots are useful. Additionally, do you have any idea of how you would accomplish your goal using plain React, or even plain HTML/JS? If you do, a similar approach should work with Ariakit.

> 3. What is the difference between `useValidate` and `onValidate`?

`useValidate` is the hook form of `onValidate`. It's something like this:

```jsx
function useValidate(callback) {
  useEffect(() => {
    const removeOnValidate = onValidate(callback);
    return removeOnVaidate;
  }, []);
}
```

You really shouldn't use `onValidate` directly when working with React.

**Reply by mule-stand** [reply]
Its code example using plain React

and similarly there could be a check for an empty input field.

and the input is fully controlled and the logic prevents the user from entering an invalid price


```ts
const enum AmountValidationError {
  MaxLimitExceeded = 'MAX_LIMIT_EXCEEDED',
}


const MAX_AMOUNT = 10000

const validateAmount = (amount: number): AmountValidationError | null => {
  if (amount > MAX_AMOUNT) {
    return AmountValidationError.MaxLimitExceeded
  }
  return null
}

const AmountField = () => {
  const [value, setValue] = React.useState(0)
  const [inputValue, setInputValue] = React.useState('')
  const [error, setError] = React.useState<string | null>(null)

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = event.currentTarget.value
    if (
      /^(0|[1-9]\d*)(\.\d{1,2})?$/.test(inputValue) ||
      /^(0|[1-9]\d*)\.$/.test(inputValue) ||
      inputValue === ''
    ) {
      setInputValue(inputValue)
      const parsedInputValue = parseFloat(inputValue)
      const error = validateAmount(parsedInputValue)
      if (error) {
        setError(error)
        return
      }
      setValue(parsedInputValue)
      setError(null)
    }
  }

  return (
    <div>
      <Input type="text" value={inputValue} onChange={handleInputChange} />
      {error && <div>{error}</div>}
      <div>
        <pre>{JSON.stringify(value, null, 2)}</pre>
      </div>
    </div>
  )
}

```
